### ISSUE TITLE
[css-text] How to handle leading ideographic space sequences ?

### ISSUE THREAD

**Issue by javifernandez** [w3c/csswg-drafts#4180]
I'm not sure whether the spec covers some extreme cases related to full-width spaces, specifically ideographich spaces (x3000). 

As far as I know, leading ideographich space sequences should not be collapsed, even when the white-space property allows to do so (normal, nowrap, or pre-line). However, I'm not sure what to do in cases like this one:

`<div width: 2em; white-space:normal">&#x3000;&#x3000;XX</div>`

First of all, 'white-space: normal' allows wrapping, as it's [stated](https://drafts.csswg.org/css-text-3/#valdef-white-space-normal) in the spec: 

> Lines may wrap at allowed soft wrap opportunities, as determined by the line-breaking rules in effect, in order to minimize inline-axis overflow. 

The [Phase II: Trimming and Positioning](https://drafts.csswg.org/css-text-3/#white-space-phase-2) section describes several scenarios and steps to evaluate, but I don't know which one this cases should be applied to:

> 1- A sequence of collapsible spaces at the beginning of a line (ignoring any intervening inline box boundaries) is removed. 

In this case, these are full-width spaces, so they can not be considered as collapsible, despite the value of the white-space property (normal).

> 3- A sequence at the end of a line (ignoring any intervening inline box boundaries) of collapsible spaces (U+0020) and/or other space separators whose white-space value collapses spaces is removed. 

If we allow breaking after the last U+3000 space, the space sequence will fill the whole line. Should we apply this step and remove the sequence ? In that case, it'd be the same than applying the step 1, which I think is not what we want for ideographic spaces.

So, I wonder which of the following results would be the expected one, according to the current specs, for this case:

<table>
<tr><td><b>Result 1</b></td><td><b>Result 2</b></td><td><b>Result 3</b></td></tr>
<tr><td><img src="https://user-images.githubusercontent.com/2230241/62632730-3f6ba280-b933-11e9-8e21-80ecdab3bee9.png"/></td><td><img src="https://user-images.githubusercontent.com/2230241/62632729-3f6ba280-b933-11e9-9ed5-f49338007941.png"/></td><td><img src="https://user-images.githubusercontent.com/2230241/62632728-3ed30c00-b933-11e9-9495-184e8f87a6b1.png"/></td></tr>
</table>

**Comment by frivoal** [comment]
> which of the following results would be the expected one

Result 2 is the expected one.

Here is why:

* Phase 1:
    > A sequence of collapsible spaces at the beginning of a line is removed

    Ideographic spaces are not "collapsible spaces".
    => They do not get removed
    (Nothing else is relevant in "Phase 1")
* Segment Break Transformation Rules
    (Nothing is relevant in that section)
* Phase 2:
   * > as each line is laid out

        wrapping opportunities are allowed after every ideographic space (cf: https://www.unicode.org/reports/tr14/tr14-39.html#BA)
        => We can fit the two U+3000 but no more, and are allowed to wrap after each of them, so they both go on line 1, and XX goes on line 2
    * > A sequence at the end of a line of collapsible spaces (U+0020) and/or other space separators whose white-space value collapses spaces is removed. 

        U+3000 is one of "[other space separators](https://drafts.csswg.org/css-text-3/#white-space-rules)"
        => the two U+3000 at the end of line 1 get removed. That does not change what goes on line 1 or 2, as this happens after splitting into lines. so the XX stay on line 2. However, if there was a span with a background around the U+3000, or a `text-decoration:underline`, the background or underline would not be seen on line 1, since the U+3000 are removed.
        (Nothing else is relevant in "Phase 2")

**Comment by javifernandez** [comment]
Good ! Thanks for the explanation, which confirms my understanding as well.

**Comment by javifernandez** [comment]
I'm facing some troubles implementing the handling of leading ideographic spaces, so before reopening this issue, I'd like to clarify some details of @frivoal explanation above.

1- I'm quite shocked about the statement that " the two U+3000 at the end of line 1 get removed." since this sequence is both, a leading and a trailing sequence at the same time. Are you sure we want this ambiguity to be applied on any case ? This would imply that steps 1 of 2 of Phase 2 produce the same result, even though ideographic spaces are not collapsible spaces:

https://drafts.csswg.org/css-text-3/#white-space-phase-2

> 1- A sequence of collapsible spaces at the beginning of a line (ignoring any intervening inline box boundaries) is removed. 
> ...
> 3- A sequence at the end of a line (ignoring any intervening inline box boundaries) of collapsible spaces (U+0020) and/or other space separators whose white-space value collapses spaces is removed. 

I'm not sure how to implement this behavior of wrapping the line, so that the "X" character is moved to the second line, but then removing the spaces that motivated the line wrapping. The fact that this is only perceptible if the space sequence has some specific style (background or underline) is also strange.

**Comment by frivoal** [comment]
Yes, I am quite sure that this is what the spec says, and it seems like good behavior to me.

ideographic spaces don't collapse with eachoter, or with spaces. Unlike regular spaces, they being at the beginning of the line doesn't cause them to be discarded, but they do get discarded when they are the end of the line. In the case you are talking about, they are both at the beginning and at the end of the line, because there's nothing else in the line. In that case, yes, it give the same results as step 1 even though they're not collapsible, but that's just true in this special case, not in general.

As for how to implement, how about replacing them with a preserved line feed in that situation?

If that turns out to be too hard, I suppose we could replace step 3 two steps, like:
> 3 A sequence at the end of a line (ignoring any intervening inline box boundaries) of collapsible spaces (U+0020) is removed
> 3.5 A sequence at the end of a line (ignoring any intervening inline box boundaries) of other space separators whose white-space value collapses spaces and intervening collapsible spaces (U+0020) not removed at step 3 must hang.

(that's kind of the same as the first sub bullet of step 4, for collapsible values of white-space in addition non conditional on white-space:pre-wrap. Also, note that we have an open issue on the details of step 4 https://github.com/w3c/csswg-drafts/issues/3440).

That seems more complicated to me, to largely achieve the same results, but it wouldn't leave us with empty lines (and authors could paint a background on these if they wanted).

**Comment by javifernandez** [comment]
I think we can leave the spec as it is for now; I just wanted to double check whether the behavior for this special case makes sense. I got some progress to implement it, so we can leave this issue closed.

Thanks again for the promptly feedback.

**Comment by kojiishi** [comment]
Let me re-open this so that we can discuss a bit more. I think wasn't thinking well-enough when it was resolved, sorry about that, but I agree with @javifernandez that we should hang trailing ideographic spaces, without removing them.

I have a few reasons to believe so, but I'd like to see what others think about this: @fantasai @MurakamiShinyu @himorin @r12a @macnmm

**Comment by frivoal** [comment]
> we should hang trailing ideographic spaces, without removing them.

I'd be OK with that.

**Comment by stantonma** [comment]
Hanging multiple ideographic spaces could cause the inline bounds to extend off screen though - making things like background-color and selection also go off screen.

**Comment by frivoal** [comment]
I've made PRs to enact the proposed change:
Spec PR: https://github.com/w3c/csswg-drafts/pull/4290
Test PR: https://github.com/web-platform-tests/wpt/pull/18949

**Comment by kojiishi** [comment]
@stantonma 
> Hanging multiple ideographic spaces could cause the inline bounds to extend off screen though - making things like background-color and selection also go off screen.

Yeah, that is one of the effects I would like. I think the ideographic space at line end should be selectable, just like other hanging punctuation are.

**Comment by javifernandez** [comment]
Hi florian,

Despite we have closed this issue, which I agree with the resolution, I'm not sure the changes in the test reflected what we have resolved; it may be possible I haven't understood it correctly, so I'd appreciate this get clarified before I reopen it, or maybe better create a new issue.

The test I'm talking about is [trailing-ideographic-space-001.html ](https://github.com/web-platform-tests/wpt/blob/master/css/css-text/white-space/trailing-ideographic-space-001.html). There is a comment in that test precisely stating that the trailing ideographic space, which is hanging, should not extend the width of its parent, which has max-content width. 

However, this trailing ideographic space is followed by a forced-break, so If I understood correctly what the spec dictates, it should [conditionally hang](https://drafts.csswg.org/css-text-3/#conditionally-hang), which means that it should account for it's parent's max-content width.

>  Glyphs that conditionally hang are not taken into account when computing min-content sizes and any sizes derived thereof, but they are taken into account for max-content sizes and any sizes derived thereof.
